core/server/master/src/main/java/alluxio/master/file/LostFileDetector.java

```diff
@@ -37,40 +43,65 @@ final class LostFileDetector implements HeartbeatExecutor {
   private static final Logger LOG = LoggerFactory.getLogger(LostFileDetector.class);
   private final FileSystemMaster mFileSystemMaster;
   private final InodeTree mInodeTree;
+  private final BlockMaster mBlockMaster;
 
   /**
    * Constructs a new {@link LostFileDetector}.
    */
-  public LostFileDetector(FileSystemMaster fileSystemMaster, InodeTree inodeTree) {
+  public LostFileDetector(FileSystemMaster fileSystemMaster, BlockMaster blockMaster,
+      InodeTree inodeTree) {
     mFileSystemMaster = fileSystemMaster;
     mInodeTree = inodeTree;
+    mBlockMaster = blockMaster;
     MetricsSystem.registerCachedGaugeIfAbsent(MetricKey.MASTER_LOST_FILE_COUNT.getName(),
         () -> mFileSystemMaster.getLostFiles().size());
   }
 
   @Override
   public void heartbeat() throws InterruptedException {
-    for (long fileId : mFileSystemMaster.getLostFiles()) {
+    Iterator<Long> iter = mBlockMaster.getLostBlocksIterator();
+    Set<Long> lostFiles = new HashSet<>();
+    while (iter.hasNext()) {
+      long blockId = iter.next();
+      long containerId = BlockId.getContainerId(blockId);
+      long fileId = IdUtils.createFileId(containerId);
+      lostFiles.add(fileId);
+      iter.remove();
```

Comment:
The documention of `Iterator::remove` says 

  The behavior of an iterator is unspecified if the underlying collection is modified while the iteration is in progress in any way other than by calling this method.

Since the underlying collection is `DefaultBlockMaster#mLostBlocks`, which is shared and may be mutated by the thread running `DefaultBlockMaster`, I'm concerned what would be the implications here.

Reply:
AFAIK, the `CurrentHashSet` is ThreadSafe, so I guess it can support iterate, remove and modify at the same time?

Reply:
```java
  public static void main(String[] args) {
    ConcurrentHashSet a = new ConcurrentHashSet();
    a.add(1);
    a.add(2);
    Iterator it = a.iterator();
    while(it.hasNext()) {
      Object tmp = it.next();
      it.remove();
    }
  }
```

I did a small test, I removed all the items within the loop through the debug way

Reply:
https://stackoverflow.com/a/3768659/4933827
By "the iterator is not thread safe" it means as long as you don't use the same iterator in multiple threads concurrently, you will be fine. Even if two threads are working with two iterators and doing add/remove operations, the thread safety is handled by the underlying ConcurrentHashSet. I think we are good here.